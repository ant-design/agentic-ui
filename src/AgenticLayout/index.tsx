import { ConfigProvider } from 'antd';
import { useMergedState } from 'rc-util';
import React, { ReactNode, useContext } from 'react';
import { useAgenticLayoutStyle } from './style';

export interface AgenticLayoutProps {
  /** 左侧内容 */
  left?: ReactNode;
  /** 中间内容 */
  center: ReactNode;
  /** 右侧内容 */
  right?: ReactNode;
  /** 左侧是否可折叠 */
  leftCollapsible?: boolean;
  /** 右侧是否可折叠 */
  rightCollapsible?: boolean;
  /** 左侧折叠状态（受控） */
  leftCollapsed?: boolean;
  /** 右侧折叠状态（受控） */
  rightCollapsed?: boolean;
  /** 左侧默认折叠状态（非受控时使用） */
  leftDefaultCollapsed?: boolean;
  /** 右侧默认折叠状态（非受控时使用） */
  rightDefaultCollapsed?: boolean;
  /** 左侧折叠状态变化回调 */
  onLeftCollapse?: (collapsed: boolean) => void;
  /** 右侧折叠状态变化回调 */
  onRightCollapse?: (collapsed: boolean) => void;
  /** 自定义样式 */
  style?: React.CSSProperties;
  /** 自定义类名 */
  className?: string;
  /** 左侧宽度 */
  leftWidth?: number;
  /** 右侧宽度 */
  rightWidth?: number;
  /** 最小高度 */
  minHeight?: string | number;
}

/**
 * AgenticLayout 组件 - 智能体布局组件
 *
 * 该组件提供一个三栏布局的容器，支持左中右三个区域的灵活配置。
 * 左右侧栏支持折叠功能，中间区域自适应宽度。
 *
 * @component
 * @description 智能体布局组件，提供左中右三栏布局
 * @param {AgenticLayoutProps} props - 组件属性
 * @param {ReactNode} [props.left] - 左侧内容
 * @param {ReactNode} props.center - 中间内容
 * @param {ReactNode} [props.right] - 右侧内容
 * @param {boolean} [props.leftCollapsible=true] - 左侧是否可折叠
 * @param {boolean} [props.rightCollapsible=true] - 右侧是否可折叠
 * @param {boolean} [props.leftDefaultCollapsed=false] - 左侧默认折叠状态
 * @param {boolean} [props.rightDefaultCollapsed=false] - 右侧默认折叠状态
 * @param {(collapsed: boolean) => void} [props.onLeftCollapse] - 左侧折叠状态变化回调
 * @param {(collapsed: boolean) => void} [props.onRightCollapse] - 右侧折叠状态变化回调
 * @param {React.CSSProperties} [props.style] - 自定义样式
 * @param {string} [props.className] - 自定义CSS类名
 * @param {number} [props.leftWidth=256] - 左侧宽度
 * @param {number} [props.rightWidth=256] - 右侧宽度
 * @param {string | number} [props.minHeight='600px'] - 最小高度
 *
 * @example
 * ```tsx
 * <AgenticLayout
 *   left={<History />}
 *   center={<ChatFlowContainer />}
 *   right={<Workspace />}
 *   leftCollapsible={true}
 *   rightCollapsible={true}
 *   onLeftCollapse={(collapsed) => console.log('左侧折叠:', collapsed)}
 *   onRightCollapse={(collapsed) => console.log('右侧折叠:', collapsed)}
 * />
 * ```
 *
 * @returns {React.ReactElement} 渲染的布局组件
 *
 * @remarks
 * - 支持左右侧栏的独立折叠控制
 * - 中间区域自适应剩余宽度
 * - 提供响应式布局适配
 * - 支持自定义宽度和高度
 * - 集成 Ant Design 主题系统
 */
export const AgenticLayout: React.FC<AgenticLayoutProps> = ({
  left,
  center,
  right,
  leftCollapsed: controlledLeftCollapsed,
  rightCollapsed: controlledRightCollapsed,
  leftDefaultCollapsed = false,
  rightDefaultCollapsed = false,
  onLeftCollapse,
  onRightCollapse,
  style,
  className,
  leftWidth = 256,
  rightWidth = 540,
}) => {
  const { getPrefixCls } = useContext(ConfigProvider.ConfigContext);
  const prefixCls = getPrefixCls('agentic-layout');
  const { wrapSSR, hashId } = useAgenticLayoutStyle(prefixCls);

  // 使用 useMergedState 管理左侧折叠状态
  const [leftCollapsed] = useMergedState(leftDefaultCollapsed, {
    value: controlledLeftCollapsed,
    onChange: onLeftCollapse,
  });

  // 使用 useMergedState 管理右侧折叠状态
  const [rightCollapsed] = useMergedState(rightDefaultCollapsed, {
    value: controlledRightCollapsed,
    onChange: onRightCollapse,
  });

  return wrapSSR(
    <div className={`${prefixCls} ${className || ''} ${hashId}`} style={style}>
      {/* 左侧边栏 */}
      {left && (
        <div
          className={`${prefixCls}-sidebar ${prefixCls}-sidebar-left ${
            leftCollapsed ? `${prefixCls}-sidebar-left-collapsed` : ''
          } ${hashId}`}
          style={{
            width: leftCollapsed ? 0 : leftWidth,
            minWidth: leftCollapsed ? 0 : leftWidth,
            maxWidth: leftCollapsed ? 0 : leftWidth,
          }}
        >
          <div className={`${prefixCls}-sidebar-content ${hashId}`}>{left}</div>
        </div>
      )}

      {/* 中间内容区域 */}
      <div
        className={`${prefixCls}-main ${hashId}`}
        style={{
          flex: 1,
          display: 'flex',
          minWidth: 0,
        }}
      >
        {center}
      </div>

      {/* 右侧边栏 */}
      {right && (
        <div
          className={`${prefixCls}-sidebar ${prefixCls}-sidebar-right ${
            rightCollapsed ? `${prefixCls}-sidebar-right-collapsed` : ''
          } ${hashId}`}
          style={{
            width: rightCollapsed ? 0 : rightWidth,
            minWidth: rightCollapsed ? 0 : rightWidth,
            maxWidth: rightCollapsed ? 0 : rightWidth,
          }}
        >
          <div className={`${prefixCls}-sidebar-content ${hashId}`}>
            {right}
          </div>
        </div>
      )}
    </div>,
  );
};
