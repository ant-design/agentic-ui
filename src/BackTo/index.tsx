import { BackBottom } from './BackBottom';
import { BackTop } from './BackTop';

export type { BackBottomProps } from './BackBottom';
export type { BackTopProps } from './BackTop';

export const BackTo = {
  Top: BackTop,
  Bottom: BackBottom,
};
