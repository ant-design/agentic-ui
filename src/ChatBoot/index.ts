export { default as Title } from './Title';
export type { TitleProps } from './Title';

export { default as CaseReply } from './CaseReply';
export type { CaseReplyProps } from './CaseReply';

export { default as ButtonTab } from './ButtonTab';
export type { ButtonTabProps } from './ButtonTab';

export { default as ButtonTabGroup } from './ButtonTabGroup';
export type { ButtonTabGroupProps, ButtonTabItem } from './ButtonTabGroup';
