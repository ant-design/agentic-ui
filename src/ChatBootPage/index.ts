/**
 * @file ChatBootPage 组件模块导出
 * 
 * 该文件导出聊天启动页面相关组件。
 * 所有组件的详细文档请参考各自的组件定义文件。
 */

export { default as Title } from './Title';
export type { TitleProps } from './Title';

export { default as CaseReply } from './CaseReply';
export type { CaseReplyProps } from './CaseReply';

export { default as ButtonTab } from './ButtonTab';
export type { ButtonTabProps } from './ButtonTab';

export { default as ButtonTabGroup } from './ButtonTabGroup';
export type { ButtonTabGroupProps, ButtonTabItem } from './ButtonTabGroup';
