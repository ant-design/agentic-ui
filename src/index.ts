export * from './MarkdownEditor/editor/Preview/index';
export { schemaToHeading } from './MarkdownEditor/editor/tools/Leading';
export * from './MarkdownEditor/editor/utils';
export * from './MarkdownEditor/editor/utils/docx/index';
export * from './MarkdownEditor/index';
export * from './Slides';
