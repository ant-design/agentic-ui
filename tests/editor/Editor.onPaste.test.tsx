import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { MarkdownEditor } from '../../src';

// Mock URL constructor for testing
global.URL = class URL {
  searchParams: URLSearchParams;
  constructor(url: string) {
    this.searchParams = new URLSearchParams();
    if (url.includes('?')) {
      const [, params] = url.split('?');
      this.searchParams = new URLSearchParams(params);
    }
  }
} as any;

// Mock window.getSelection
Object.defineProperty(window, 'getSelection', {
  writable: true,
  value: vi.fn(() => ({
    rangeCount: 0,
    getRangeAt: vi.fn(),
    removeAllRanges: vi.fn(),
    anchorNode: null,
    focusNode: null,
    type: 'None',
  })),
});

describe('Editor onPaste function', () => {
  let mockEditor: any;
  let mockStore: any;
  let container: HTMLElement;

  beforeEach(() => {
    vi.clearAllMocks();
    mockEditor = {
      children: [{ type: 'paragraph', children: [{ text: '' }] }],
      selection: {
        anchor: { path: [0, 0], offset: 0 },
        focus: { path: [0, 0], offset: 0 },
      },
      operations: [],
    };

    // Mock store
    mockStore = {
      insertLink: vi.fn(),
    };

    // Setup DOM
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    document.body.removeChild(container);
  });

  const createClipboardEvent = (
    data: Record<string, string>,
    files?: File[],
    target?: HTMLElement,
  ) => {
    const clipboardData = {
      types: Object.keys(data),
      getData: vi.fn((type: string) => data[type] || ''),
      files: files || [],
    };

    return {
      clipboardData,
      preventDefault: vi.fn(),
      stopPropagation: vi.fn(),
      target: target || container,
      isTrusted: true,
      isComposing: false,
      nativeEvent: {
        clipboardData,
      },
    } as any;
  };

  // Add a basic test to verify the editor renders and can receive paste events
  it('should render editor and handle basic paste event', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;
    expect(editableElement).toBeTruthy();

    // Wait for the editor to be fully initialized
    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    // Create a simple paste event with the correct target
    const event = createClipboardEvent(
      {
        'text/plain': 'Hello World',
      },
      [],
      editableElement,
    );

    // Fire the paste event - this should not throw an error
    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();

    // Verify that the editor is still functional after the paste event
    expect(editableElement).toBeInTheDocument();
    expect(editableElement.getAttribute('contenteditable')).toBe('true');

    await waitFor(() => {
      expect(screen.getByText('Hello World')).toBeInTheDocument();
    });
    expect(editableElement).matchSnapshot();
  });

  it('should handle pasting Slate markdown fragment', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const slateFragment = JSON.stringify([
      {
        type: 'paragraph',
        children: [{ text: 'Pasted from Slate' }],
      },
    ]);

    const event = createClipboardEvent(
      {
        'application/x-slate-md-fragment': slateFragment,
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    await waitFor(() => {
      expect(screen.getByText('Pasted from Slate')).toBeInTheDocument();
    });
  });

  it('should handle pasting HTML content', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const htmlContent =
      '<p><strong>Bold text</strong> and <em>italic text</em></p>';

    const event = createClipboardEvent(
      {
        'text/html': htmlContent,
        'text/plain': 'Bold text and italic text',
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    await waitFor(() => {
      expect(
        screen.getByText((content, element) => {
          return content.includes('Bold text and');
        }),
      ).toBeInTheDocument();
    });
    await waitFor(() => {
      expect(
        screen.getByText((content, element) => {
          return content.includes('italic text');
        }),
      ).toBeInTheDocument();
    });
  });

  it('should handle pasting image files', async () => {
    const mockUpload = vi
      .fn()
      .mockResolvedValue('https://example.com/image.jpg');

    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
        image={{ upload: mockUpload }}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const imageFile = new File(['image content'], 'test.jpg', {
      type: 'image/jpeg',
    });

    const event = createClipboardEvent(
      {
        Files: '',
      },
      [imageFile],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    await waitFor(() => {
      expect(mockUpload).toHaveBeenCalledWith([imageFile]);
    });
  });

  it('should handle pasting URLs as links', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const event = createClipboardEvent(
      {
        'text/plain': 'https://example.com',
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    // The URL should be handled as a link
    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });

  it('should handle pasting image URLs', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const event = createClipboardEvent(
      {
        'text/plain': 'https://example.com/image.jpg',
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });

  it('should handle pasting media:// URLs', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const event = createClipboardEvent(
      {
        'text/plain': 'media://test?url=https://example.com/image.jpg',
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });

  it('should handle pasting attach:// URLs', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const event = createClipboardEvent(
      {
        'text/plain':
          'attach://test?url=https://example.com/file.pdf&name=document.pdf&size=1024',
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });

  it('should handle pasting markdown content', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const markdownContent =
      '# Heading\n\n**Bold text** and *italic text*\n\n- List item 1\n- List item 2';

    const event = createClipboardEvent(
      {
        'text/plain': markdownContent,
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });

  it('should handle pasting when clipboard data is undefined', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    // Create event with undefined clipboardData
    const event = {
      clipboardData: undefined,
      preventDefault: vi.fn(),
      stopPropagation: vi.fn(),
      target: editableElement,
    } as any;

    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });

  it('should handle pasting empty content', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const event = createClipboardEvent(
      {
        'text/plain': '',
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });

  it('should handle pasting with invalid JSON in slate fragment', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const event = createClipboardEvent(
      {
        'application/x-slate-md-fragment': 'invalid json',
        'text/plain': 'fallback text',
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    await waitFor(() => {
      expect(
        screen.getByText((content, element) => {
          return content.includes('fallback text');
        }),
      ).toBeInTheDocument();
    });
  });

  it('should handle pasting card elements with proper structure', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const cardFragment = JSON.stringify([
      {
        type: 'card',
        children: [{ text: 'Card content' }],
      },
    ]);

    const event = createClipboardEvent(
      {
        'application/x-slate-md-fragment': cardFragment,
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });

  it('should handle pasting in text area mode', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
        textAreaProps={{ enable: true }}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const slateFragment = JSON.stringify([
      {
        type: 'paragraph',
        children: [{ text: 'Simple text' }],
      },
    ]);

    const event = createClipboardEvent(
      {
        'application/x-slate-md-fragment': slateFragment,
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    await waitFor(() => {
      expect(screen.getByText('Simple text')).toBeInTheDocument();
    });
  });

  it('should handle pasting RTF content along with HTML', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const htmlContent = '<p>Rich text content</p>';
    const rtfContent = '{\\rtf1\\ansi\\deff0 Rich text content}';

    const event = createClipboardEvent(
      {
        'text/html': htmlContent,
        'text/rtf': rtfContent,
        'text/plain': 'Rich text content',
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    await waitFor(() => {
      expect(screen.getByText('Rich text content')).toBeInTheDocument();
    });
  });

  it('should handle pasting multiple files', async () => {
    const mockUpload = vi
      .fn()
      .mockResolvedValueOnce('https://example.com/image1.jpg')
      .mockResolvedValueOnce('https://example.com/image2.jpg');

    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
        image={{ upload: mockUpload }}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const file1 = new File(['image1'], 'test1.jpg', { type: 'image/jpeg' });
    const file2 = new File(['image2'], 'test2.jpg', { type: 'image/jpeg' });

    const event = createClipboardEvent(
      {
        Files: '',
      },
      [file1, file2],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    await waitFor(() => {
      expect(mockUpload).toHaveBeenCalledTimes(2);
      expect(mockUpload).toHaveBeenNthCalledWith(1, [file1]);
      expect(mockUpload).toHaveBeenNthCalledWith(2, [file2]);
    });
  });

  it('should handle file upload failure gracefully', async () => {
    const mockUpload = vi.fn().mockRejectedValue(new Error('Upload failed'));

    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
        image={{ upload: mockUpload }}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const imageFile = new File(['image content'], 'test.jpg', {
      type: 'image/jpeg',
    });

    const event = createClipboardEvent(
      {
        Files: '',
      },
      [imageFile],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    await waitFor(() => {
      expect(mockUpload).toHaveBeenCalledWith([imageFile]);
    });

    // Should not throw error even if upload fails
    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });

  it('should handle pasting in table cells', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[
          {
            type: 'table',
            children: [
              {
                type: 'table-row',
                children: [
                  {
                    type: 'table-cell',
                    children: [{ type: 'paragraph', children: [{ text: '' }] }],
                  },
                ],
              },
            ],
          },
        ]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const event = createClipboardEvent(
      {
        'text/plain': 'Cell content',
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    await waitFor(() => {
      expect(screen.getByText('Cell content')).toBeInTheDocument();
    });
  });

  it('should handle pasting in code blocks', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[
          {
            type: 'code',
            value: '',
            children: [{ text: '' }],
          },
        ]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const event = createClipboardEvent(
      {
        'text/plain': 'console.log("Hello World");',
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    await waitFor(() => {
      expect(
        screen.getByText('console.log("Hello World");'),
      ).toBeInTheDocument();
    });
  });

  it('should handle pasting video URLs', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const event = createClipboardEvent(
      {
        'text/plain': 'https://example.com/video.mp4',
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });

  it('should handle pasting audio URLs', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const event = createClipboardEvent(
      {
        'text/plain': 'https://example.com/audio.mp3',
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });

  it('should handle pasting malformed URLs gracefully', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const event = createClipboardEvent(
      {
        'text/plain': 'media://malformed-url-without-params',
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });

  it('should handle pasting when no selection exists', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[
          { type: 'paragraph', children: [{ text: 'Existing content' }] },
        ]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const event = createClipboardEvent(
      {
        'text/plain': 'New content',
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });

  it('should handle pasting complex markdown with multiple elements', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="true"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const complexMarkdown = `# Main Title

## Subtitle

This is a paragraph with **bold** and *italic* text.

### Code Example

\`\`\`javascript
function hello() {
  console.log("Hello World");
}
\`\`\`

### List Items

- Item 1
- Item 2
  - Nested item
- Item 3

### Links and Images

[Link text](https://example.com)

![Alt text](https://example.com/image.jpg)

### Table

| Column 1 | Column 2 |
|----------|----------|
| Cell 1   | Cell 2   |
| Cell 3   | Cell 4   |

> This is a blockquote
> with multiple lines`;

    const event = createClipboardEvent(
      {
        'text/plain': complexMarkdown,
      },
      [],
      editableElement,
    );

    fireEvent.paste(editableElement, event);

    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });

  it('should handle pasting when editor is in readonly mode', async () => {
    const { container: editorContainer } = render(
      <MarkdownEditor
        initSchemaValue={[{ type: 'paragraph', children: [{ text: '' }] }]}
        readonly={true}
      />,
    );

    const editableElement = editorContainer.querySelector(
      '[contenteditable="false"]',
    ) as HTMLElement;

    await waitFor(() => {
      expect(editableElement).toBeInTheDocument();
    });

    const event = createClipboardEvent(
      {
        'text/plain': 'Should not paste',
      },
      [],
      editableElement,
    );

    expect(() => {
      fireEvent.paste(editableElement, event);
    }).not.toThrow();
  });
});
