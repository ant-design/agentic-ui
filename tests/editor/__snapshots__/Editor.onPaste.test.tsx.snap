// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Editor onPaste function > should render editor and handle basic paste event 1`] = `
<div
  aria-multiline="true"
  autocapitalize="false"
  autocorrect="false"
  class="ant-md-editor-content-focus ant-md-editor-content ant-md-editor-content-edit css-dev-only-do-not-override-1m2bkf9"
  contenteditable="true"
  data-slate-editor="true"
  data-slate-node="value"
  role="textbox"
  spellcheck="false"
  style="position: relative; white-space: pre-wrap; word-wrap: break-word; font-size: 14px;"
  zindex="-1"
>
  <div
    class="ant-md-editor-drag-el empty"
    data-be="paragraph"
    data-slate-node="element"
    data-slate-placeholder="请输入内容..."
  >
    <span
      data-slate-node="text"
    >
      <span
        class="css-dev-only-do-not-override-1m2bkf9  css-dev-only-do-not-override-1m2bkf9"
        data-be="text"
        data-slate-leaf="true"
        draggable="false"
      >
        <span
          data-slate-length="0"
          data-slate-zero-width="n"
        >
          ﻿
          <br />
        </span>
      </span>
    </span>
    <span
      contenteditable="false"
      style="font-size: 0px; line-height: 0; display: inline-block;"
    >
       
    </span>
  </div>
  <div
    class="ant-md-editor-drag-el"
    data-be="paragraph"
    data-slate-node="element"
  >
    <span
      data-slate-node="text"
    >
      <span
        class="css-dev-only-do-not-override-1m2bkf9  css-dev-only-do-not-override-1m2bkf9"
        data-be="text"
        data-slate-leaf="true"
        draggable="false"
      >
        <span
          data-slate-string="true"
        >
          Hello World
        </span>
      </span>
    </span>
  </div>
</div>
`;
